ipa-core/src/protocol/ipa_prf/malicious_security/prover.rs

```diff
@@ -14,15 +15,64 @@ use crate::{
     },
 };
 
-pub struct ZeroKnowledgeProof<F: PrimeField, N: ArrayLength> {
-    pub g: GenericArray<F, N>,
+pub type UVPolynomial<F, N> = (GenericArray<F, N>, GenericArray<F, N>);
+pub type TwoNMinusOne<R> = Diff<Sum<R, R>, U1>;
+
+// todo: deprecate
+pub type TwoNPlusOne<R> = Sum<Sum<R, R>, U1>;
+
+/// The purpose of this trait is to define an associated type,
+/// i.e. the length of the array `Length`
+pub trait ProofArray {
+    type Length: ArrayLength;
+}
+
+/// This struct contains a single `ZeroKnowledgeProof`
+/// The length of the proof is not the generic parameter `R`.
+/// The length can be obtained using the trait `ProofArray`
+/// where the associated type `Length`
+/// represents the length of the proof.
+///
+/// Parameter `R` represents the recursion factor, which is a parameter
+/// used during the proof generation and verification.
+#[derive(Debug)]
+pub struct ZeroKnowledgeProof<F, R>
+where
+    Self: ProofArray,
+    F: PrimeField,
+    <Self as ProofArray>::Length: ArrayLength,
+{
+    pub g: GenericArray<F, <Self as ProofArray>::Length>,
+}
+
+impl<F, R> ProofArray for ZeroKnowledgeProof<F, R>
+where
+    F: PrimeField,
+    R: Add + ArrayLength,
+    <R as Add>::Output: Sub<U1>,
```

Comment:
can you explain why you need these bounds in a comment?

Reply:
Yes, there is a recursion parameter `R`. From this parameter, you can compute the length of the proof as `2R-1`, `R: Add` means `2R`. Almost any function needs to take recursion parameter `R` so it makes sense to parameterize the proofs also with this parameter rather than the actual length. Using the actual length would lead to a lot of ugly typenum arithemtics in a lot of places. This is why I have changed it.

I use this trait to clean up the trait bounds. Without the ProofArray trait, I would need to include these bounds everywhere I currently include the ProofArray trait bound.